packages/x-tree-view/src/internals/useTreeView/useTreeView.utils.ts

```diff
@@ -51,16 +84,142 @@ export const getNextItem = (
+
+  return itemId!;
 };
 
-export const getFirstItem = (instance: TreeViewInstance<[UseTreeViewItemsSignature]>) =>
-  instance.getNavigableChildrenIds(null)[0];
+export const getFirstNavigableItem = (instance: TreeViewInstance<[UseTreeViewItemsSignature]>) =>
+  instance.getChildrenIds(null).find(instance.isItemNavigable)!;
+
+/**
+ * This is used to determine the start and end of a selection range so
+ * we can get the items between the two border items.
+ *
+ * It finds the items' common ancestor using
+ * a naive implementation of a lowest common ancestor algorithm
+ * (https://en.wikipedia.org/wiki/Lowest_common_ancestor).
+ * Then compares the ancestor's 2 children that are ancestors of itemA and ItemB
+ * so we can compare their indexes to work out which item comes first in a depth first search.
+ * (https://en.wikipedia.org/wiki/Depth-first_search)
+ *
+ * Another way to put it is which item is shallower in a trémaux tree
+ * https://en.wikipedia.org/wiki/Tr%C3%A9maux_tree
+ */
+const findOrderInTremauxTree = (
+  instance: TreeViewInstance<[UseTreeViewItemsSignature]>,
+  itemAId: string,
+  itemBId: string,
+) => {
+  if (itemAId === itemBId) {
+    return [itemAId, itemBId];
+  }
+
+  const itemMetaA = instance.getItemMeta(itemAId);
+  const itemMetaB = instance.getItemMeta(itemBId);
+
+  if (itemMetaA.parentId === itemMetaB.id || itemMetaB.parentId === itemMetaA.id) {
+    return itemMetaB.parentId === itemMetaA.id
+      ? [itemMetaA.id, itemMetaB.id]
+      : [itemMetaB.id, itemMetaA.id];
+  }
+
+  const aFamily: (string | null)[] = [itemMetaA.id];
+  const bFamily: (string | null)[] = [itemMetaB.id];
+
+  let aAncestor = itemMetaA.parentId;
+  let bAncestor = itemMetaB.parentId;
+
+  let aAncestorIsCommon = bFamily.indexOf(aAncestor) !== -1;
+  let bAncestorIsCommon = aFamily.indexOf(bAncestor) !== -1;
+
+  let continueA = true;
+  let continueB = true;
+
+  while (!bAncestorIsCommon && !aAncestorIsCommon) {
+    if (continueA) {
+      aFamily.push(aAncestor);
+      aAncestorIsCommon = bFamily.indexOf(aAncestor) !== -1;
+      continueA = aAncestor !== null;
+      if (!aAncestorIsCommon && continueA) {
+        aAncestor = instance.getItemMeta(aAncestor!).parentId;
+      }
+    }
+
+    if (continueB && !aAncestorIsCommon) {
+      bFamily.push(bAncestor);
+      bAncestorIsCommon = aFamily.indexOf(bAncestor) !== -1;
+      continueB = bAncestor !== null;
+      if (!bAncestorIsCommon && continueB) {
+        bAncestor = instance.getItemMeta(bAncestor!).parentId;
+      }
+    }
+  }
+
+  const commonAncestor = aAncestorIsCommon ? aAncestor : bAncestor;
+  const ancestorFamily = instance.getChildrenIds(commonAncestor);
+
+  const aSide = aFamily[aFamily.indexOf(commonAncestor) - 1];
+  const bSide = bFamily[bFamily.indexOf(commonAncestor) - 1];
+
+  return ancestorFamily.indexOf(aSide!) < ancestorFamily.indexOf(bSide!)
+    ? [itemAId, itemBId]
+    : [itemBId, itemAId];
+};
+
+export const getNavigableItemsInRange = (
```

Comment:
I no longer use `getNextNavigableItem` here because it is using `siblings.indexOf(itemMeta.id)` whenever the current item is a leaf (which should be the case for the majority of items).
Maybe in the future we will have an easier way to access the index of an element, but for now it's a fairly expensive operation that we can avoid here since we already know the index of the item just before.

Reply:
I added an index lookup in #12213, I will extract it into a standalone PR